clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java

```diff
@@ -168,38 +179,38 @@ private static FilterResult filterTo(TopicPartition partition, Iterable<MutableR
+            boolean containsTombstones = false;
             List<Record> retainedRecords = new ArrayList<>();
 
-            try (final CloseableIterator<Record> iterator = batch.streamingIterator(decompressionBufferSupplier)) {
-                while (iterator.hasNext()) {
-                    Record record = iterator.next();
-                    filterResult.messagesRead += 1;
-
-                    if (filter.shouldRetainRecord(batch, record)) {
-                        // Check for log corruption due to KAFKA-4298. If we find it, make sure that we overwrite
-                        // the corrupted batch with correct data.
-                        if (!record.hasMagic(batchMagic))
-                            writeOriginalBatch = false;
-
-                        if (record.offset() > maxOffset)
-                            maxOffset = record.offset();
-
-                        retainedRecords.add(record);
-                    } else {
-                        writeOriginalBatch = false;
-                    }
-                }
-            }
+            final BatchIterationResult iterationResult = iterateOverBatch(batch, decompressionBufferSupplier, filterResult, filter,
+                                                                          batchMagic, writeOriginalBatch, maxOffset, retainedRecords);
+            containsTombstones = iterationResult.containsTombstones();
+            writeOriginalBatch = iterationResult.shouldWriteOriginalBatch();
+            maxOffset = iterationResult.maxOffset();
 
             if (!retainedRecords.isEmpty()) {
-                if (writeOriginalBatch) {
+                // we check if the delete horizon should be set to a new value
+                // in which case, we need to reset the base timestamp and overwrite the timestamp deltas
+                // if the batch does not contain tombstones, then we don't need to overwrite batch
+                boolean needToSetDeleteHorizon = batch.magic() >= 2 && (containsTombstones || containsEmptyMarker)
+                    && !batch.deleteHorizonSet();
+                if (writeOriginalBatch && (!needToSetDeleteHorizon)) {
                     batch.writeTo(bufferOutputStream);
                     filterResult.updateRetainedBatchMetadata(batch, retainedRecords.size(), false);
                 } else {
-                    MemoryRecordsBuilder builder = buildRetainedRecordsInto(batch, retainedRecords, bufferOutputStream);
+                    final MemoryRecordsBuilder builder;
+                    if (needToSetDeleteHorizon) {
+                        long deleteHorizonMs = retrieveDeleteHorizon(batch, filter, containsEmptyMarker);
+                        builder = buildRetainedRecordsInto(batch, retainedRecords, bufferOutputStream, deleteHorizonMs);
+                        if (deleteHorizonMs > filterResult.latestDeleteHorizon()) {
+                            filterResult.updateLatestDeleteHorizon(deleteHorizonMs);
```

Comment:
This doesn't look right. We need to track not only newly generated deletionHorizon but also existing one if the batch is kept.

Reply:
@junrao Aren't we already keeping track of each individual delete horizon in each batch's first timestamp? My impression was that this result would just return the biggest delete horizon seen so far.

Reply:
@ConcurrencyPractitioner : If we get into the else branch in line 210, it seems that we still need to call filterResult.updateLatestDeleteHorizon() since the batch may contain deleteHorizon?

Reply:
Oh, I see. Makes sense. I misunderstood what the comment was suggesting.

Reply:
Just a note, I actually did resolve this comment with my previous push. Turns out I spotted this error while running over the code previously. Just didn't realize that it actually resolved this one as well.